pylib/gyp/msvs_emulation.py

```diff
@@ -978,14 +974,70 @@ def GenerateEnvironmentFiles(toplevel_build_dir, generator_flags, system_include
     f.write(env_block)
     f.close()
 
-    # Find cl.exe location for this architecture.
-    args = vs.SetupScript(arch)
-    args.extend(('&&', 'for', '%i', 'in', '(cl.exe)', 'do', '@echo', 'LOC:%~$PATH:i'))
-    output = subprocess.check_output(args, shell=True).decode('utf-8')
-    cl_paths[arch] = _ExtractCLPath(output)
+    cl_paths[arch] = env['GYP_CL_PATH']
   return cl_paths
 
 
+def _GetEnvironment(arch, vs, open_out):
+  """
+  This function will run the VC environment setup script, retrieve variables,
+  and also the path on cl.exe.
+  It will then try to cache the values to disk, and on next run will try to
+  lookup the cache. The cache key is the path to the setup script (which is
+  embedded within each Visual Studio installed instance) + it's args.
+  Even after a cache hit we do some validation of the cached values,
+  since parts of the tool-set can be upgraded with in the installed lifecycle
+  so paths and version numbers may change.
+
+  Args:
+    arch: {string} target architecture
+    vs: VisualStudioVersion
+    open_out: file open wrapper
+
+  Returns: {dict} the important environment variables VC need to run
+
+  """
+  args = vs.SetupScript(arch)
+  args.extend(('&&', 'set', '&&', 'where', 'cl.exe'))
+  args_slug = ''.join(args).encode('utf-8')
+  args_hash = hashlib.md5(args_slug).hexdigest()
+  cache_key = 'gyp-env-cache-' + args_hash
+  # The default value for %TEMP% will make all cache look ups to safely miss
+  cache_dir = os.environ.get('GYP_TEMP', os.getcwd())
+  cache_keyed_file = os.path.join(cache_dir, cache_key)
+  if os.path.exists(cache_keyed_file):
+    try:
+      with open(cache_keyed_file, 'rb') as f:
+        env = pickle.load(f)
+    except Exception as e:
+      DebugOutput(DEBUG_GENERAL, "Failed to load env pickle: %s" % e)
+      DebugOutput(DEBUG_GENERAL, "args_slug=%s" % args_slug)
+      DebugOutput(DEBUG_GENERAL, "cache_keyed_file=%s" % cache_keyed_file)
+      DebugOutput(DEBUG_GENERAL, traceback.format_exc())
+    cl_path = env.get('GYP_CL_PATH', '')
+    if os.path.exists(cl_path):
+      return env
+    else:
+      # cache has become invalid (probably form a tool set update)
+      os.remove(cache_keyed_file)
+  start_time = time.clock()
+  # Extract environment variables for subprocesses.
+  std_out = subprocess.check_output(args, shell=True, stderr=subprocess.STDOUT).decode('utf-8')
+  end_time = time.clock()
+  if DEBUG_GENERAL in gyp.debug.keys():
```

Comment:
__.keys()__ is not required.

Reply:
Ack.
Any lint rule we can turn on for such things?

Reply:
Nope.  Adding the __.keys()__ is legal but not necessary.